### DISCUSSION TITLE
Prevent Pipeline from rebuilding shared parent stage in multi-stage build

### DISCUSSION THREAD

**Discussion by DrewVartanian** [aws/copilot-cli#3599] [Q&A] [answered]
Please excuse any incorrect terminology, as I'm pretty new to both Copilot and docker in general.  I have multiple services that have a shared parent stage in a multi-stage docker build. Something akin to creating web and worker images like shown here:
```
FROM ubuntu:18.04 AS base
...

FROM base AS web
...

FROM base AS worker
...
```
When I deploy with Pipeline set up through Copilot, it builds both web and worker all the way from ubuntu:18.04, rather than just building base once for the first image, and then reusing it for the second.  Is there any way to have Pipeline only build base once?  Base dose take in ARG(s), but they are set to the same value in the web and worker manifest (not sure if that affects things).

**Comment by dannyrandall** [comment, marked as answer]
Hey @DrewVartanian! Using `ARG` in docker build makes it so future instructions are not cached. Some details around that are available [here](https://docs.docker.com/engine/reference/builder/#impact-on-build-caching). 

One potential solution you could try is moving the `ARG` command into the `web` and `worker` stages like this:
```
FROM base AS web
ARG name

FROM base AS worker
ARG name
```

This way, the `base` stage layers will be cache-able and should only run once in your pipeline. Let me know if this works for you!:blush:

### EDIT
Another potential solution: https://github.com/aws/copilot-cli/issues/3510#issuecomment-1136317004

**Comment by DrewVartanian** [comment]
Hi @dannyrandall .  Thanks for the response!  I tried replacing my ARGs with hardcoded ENVs to test, as they are used in the base stage build, but it still seems to be rebuilding base for each container.  Is there anything else that prevents the base layer from being cache-able (like a COPY)?

**Reply by dannyrandall** [reply]
`COPY` is [cache-able](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache), as long as the contents that are copied into the container are the same - meaning the files copied into the base stage build are the same on each build. 

Just to check, I'm assuming you're using the `target` field in [`image.build`](https://aws.github.io/copilot-cli/docs/manifest/lb-web-service/#image-build), right? And setting the target for one service to `web` and the other to `worker`?

Also, is the container you're building a linux container?

**Reply by DrewVartanian** [reply]
That's correct.  I haven't changed anything in the buildspec.yml that was generated by copilot when setting up the pipeline, and I'm using the dockerfile, context, and target fields in my manifests' image.build section (with the targets pointing to web and worker).

**Reply by dannyrandall** [reply]
Hmm...interesting.🤔  Could you share more of your Dockerfile and the `image` configuration for both services in your manifests?

**Reply by dannyrandall** [reply]
Actually - if you're copying the entire context directory (`COPY . <dest>`), I wonder if this would help? https://github.com/aws/copilot-cli/issues/3510#issuecomment-1136317004

**Reply by DrewVartanian** [reply]
I am copying the entire context directory!  Good catch!  I added the default output dir`/infrastructure/` to .dockerignore, and it's now caching properly.  Do you know if there is some way to set the output-dir in the copilot manifests?  Either way, thanks for the help!

**Reply by dannyrandall** [reply]
Yay!:tada: There isn't a way to specify the output-dir from the manifest, but you can modify the buildspec (default lines [59](https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/template/templates/cicd/buildspec.yml#L59) and [66](https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/template/templates/cicd/buildspec.yml#L66)) and specify the output directory there.
